contracts/mocks/MockMarginCalculator.sol

```diff
@@ -14,6 +15,61 @@ contract MockMarginCalculator {
         view
         returns (bool isValid)
     {
+        if (finalVault.shortOtokens.length == 0 || finalVault.shortAmounts.length == 0) return true;
+
+        bool isExcess = getExcessMargin(finalVault, OtokenInterface(finalVault.shortOtokens[0]).collateralAsset());
+
+        return isExcess;
+    }
+
+    function getExcessMargin(MarginAccount.Vault memory _vault, address _denominated)
+        public
+        view
+        returns (bool isExcess)
+    {
+        // ensure the number of collateral, long and short array is valid.
+        _checkIsValidSpread(_vault);
+        // ensure the long asset is valid for the short asset.
+        _checkIsMarginableLong(_vault);
+
         return true;
     }
+
+    function _checkIsValidSpread(MarginAccount.Vault memory _vault) internal pure {
+        require(_vault.shortOtokens.length <= 1, "MarginCalculator: Too many short otokens in the vault.");
+        require(_vault.longOtokens.length <= 1, "MarginCalculator: Too many long otokens in the vault.");
+        require(_vault.collateralAssets.length <= 1, "MarginCalculator: Too many collateral assets in the vault.");
+
+        require(
+            _vault.shortOtokens.length == _vault.shortAmounts.length,
+            "MarginCalculator: Short asset and amount mismatch."
+        );
+        require(
+            _vault.longOtokens.length == _vault.longAmounts.length,
+            "MarginCalculator: Long asset and amount mismatch."
+        );
+        require(
+            _vault.collateralAssets.length == _vault.collateralAmounts.length,
+            "MarginCalculator: Collateral asset and amount mismatch."
+        );
+    }
+
+    function _checkIsMarginableLong(MarginAccount.Vault memory _vault) internal view {
+        if (_vault.longOtokens.length == 0 || _vault.shortOtokens.length == 0) return;
+
+        OtokenInterface long = OtokenInterface(_vault.longOtokens[0]);
+        OtokenInterface short = OtokenInterface(_vault.shortOtokens[0]);
+        require(_getBatchId(long) == _getBatchId(short), "MarginCalculator: Long and short batch mismatch");
+    }
+
+    function _getBatchId(OtokenInterface _otoken) internal view returns (bytes32 id) {
+        id = keccak256(
+            abi.encodePacked(
+                _otoken.underlyingAsset(),
+                _otoken.strikeAsset(),
+                _otoken.collateralAsset(),
+                _otoken.expiryTimestamp()
+            )
+        );
+    }
```

Comment:
Move to controller
Remove from mock

Reply:
Will keep `_checkIsMarginableLong` in calculator, will move only function that checks storage state after MarginAccount lib functions.